File: taurus-protect-sdk-typescript/src/mappers/earn.ts
/**
 * Earn mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { TgvalidatordEarnReward } from '../internal/openapi/models/TgvalidatordEarnReward';
import type { EarnReward } from '../models/earn';
import { safeMap } from './base';

/**
 * Maps a reward row to an EarnReward domain model.
 */
export function earnRewardFromDto(
  dto: TgvalidatordEarnReward | null | undefined
): EarnReward | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }
  const merkl = dto.merklTokenReward;
  return {
    id: dto.id,
    recipientAddressId: dto.recipientAddressId,
    recipientAddress: dto.recipientAddress,
    rewardType: dto.rewardType,
    merklTokenReward: merkl
      ? {
          amount: merkl.amount,
          claimed: merkl.claimed,
          pending: merkl.pending,
          token: merkl.token
            ? {
                address: merkl.token.address,
                symbol: merkl.token.symbol,
                assetId: merkl.token.assetId,
              }
            : undefined,
        }
      : undefined,
  };
}

/**
 * Maps reward rows to EarnReward domain models.
 */
export function earnRewardsFromDto(
  dtos: TgvalidatordEarnReward[] | null | undefined
): EarnReward[] {
  return safeMap(dtos, earnRewardFromDto);
}

File: taurus-protect-sdk-go/pkg/protect/mapper/earn.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// EarnRewardFromDTO converts an OpenAPI EarnReward to a domain EarnReward.
func EarnRewardFromDTO(dto *openapi.TgvalidatordEarnReward) *model.EarnReward {
	if dto == nil {
		return nil
	}
	reward := &model.EarnReward{
		ID:                 safeString(dto.Id),
		RecipientAddressID: safeString(dto.RecipientAddressId),
		RecipientAddress:   safeString(dto.RecipientAddress),
	}
	if dto.RewardType != nil {
		reward.RewardType = string(*dto.RewardType)
	}
	if m := dto.MerklTokenReward; m != nil {
		merkl := &model.MerklTokenReward{
			Amount:  safeString(m.Amount),
			Claimed: safeString(m.Claimed),
			Pending: safeString(m.Pending),
		}
		if t := m.Token; t != nil {
			merkl.Token = &model.EarnRewardToken{
				Address: safeString(t.Address),
				Symbol:  safeString(t.Symbol),
				AssetID: safeString(t.AssetId),
			}
		}
		reward.MerklToken = merkl
	}
	return reward
}

// EarnRewardsFromDTO converts a page of OpenAPI EarnRewards.
func EarnRewardsFromDTO(dtos []openapi.TgvalidatordEarnReward) []*model.EarnReward {
	if dtos == nil {
		return nil
	}
	rewards := make([]*model.EarnReward, len(dtos))
	for i := range dtos {
		rewards[i] = EarnRewardFromDTO(&dtos[i])
	}
	return rewards
}
